### PR TITLE
Add binary insertion sort algorithm

### PR DESCRIPTION
**Pull request by testourii**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: sorts/binary_insertion_sort.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+For doctests run following command:
+python3 -m doctest -v binary_insertion_sort.py
+
+For manual testing run:
+python3 binary_insertion_sort.py
+"""
+
+
+def binary_search(collection: list, val: int, start: int, end: int) -> int:
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:17]: Provide the container element type as well: `list[<type>]`

File: sorts/binary_insertion_sort.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    A pure Python implementation of the binary search  algorithm.
+
+    :param collection: some ascending sorted collection with comparable items
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:21]: The docstring is a direct copy paste from `searches/binary_search.py`. Please update it as per the function.

File: sorts/binary_insertion_sort.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    elif collection[mid] > val:
+        return binary_search(collection, val, start, mid - 1)
+    else:
+        return mid
+
+
+def binary_insertion_sort(collection: list) -> list:
+    """A pure Python implementation of the binary insertion sort algorithm
+
+    :param collection: some mutable ordered collection with heterogeneous
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:55]: The docstring is a direct copy paste from `sorts/insertion_sort.py`. Please update it as per the function.

File: sorts/binary_insertion_sort.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    mid = (start + end) // 2
+    if collection[mid] < val:
+        return binary_search(collection, val, mid + 1, end)
+    elif collection[mid] > val:
+        return binary_search(collection, val, start, mid - 1)
+    else:
+        return mid
+
+
+def binary_insertion_sort(collection: list) -> list:
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:52]: Provide the container element type as well: `list[<type>]`

File: sorts/binary_insertion_sort.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    >>> collection = random.sample(range(-50, 50), 100)
+    >>> binary_insertion_sort(collection) == sorted(collection)
+    True
+    >>> import string
+    >>> collection = random.choices(string.ascii_letters + string.digits, k=100)
+    >>> binary_insertion_sort(collection) == sorted(collection)
+    True
+    """
+    for index in range(1, len(collection)):
+        val = collection[index]
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:78]: Using `enumerate` would be much more efficient here

File: sorts/binary_insertion_sort.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        mid = binary_search(collection, val, 0, index - 1)
+        collection = (
+            collection[:mid] + [val] + collection[mid:index] + collection[index + 1 :]
+        )
+    return collection
+
+
+if __name__ == "__main__":
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:86]: It's better to provide a simple CLI interface using `argparse` than to run the tests. They can be run using the `python -m doctest` command as well.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
Hey, thanks for taking your time to contribute! The documentation needs to be updated to reflect the new algorithm. There are some typing improvements and optimizations. I'll take a second look at the implemented algorithm after these changes are incorporated, but this is a good start!

**Comment by testourii** [comment]
@dhruvmanila I think I fixed the technical and conventional errors on this one. Could you take a look?

File: sorts/binary_insertion_sort.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+For doctests run following command:
+python3 -m doctest -v binary_insertion_sort.py
+
+For manual testing run:
+python3 binary_insertion_sort.py
+"""
+
+
+def binary_search(collection: list[int], value: int, start: int, end: int) -> int:
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:17]: This algorithm is incorrect and breaks with the following initial tests:
```python
binary_search([1, 2, 3, 4, 5], 2, 0, 1)  # gives 2
binary_search([1, 2, 3, 4, 5], 2, 0, 4)  # gives 2
```

Please test out the algorithms thoroughly. Also, if possible, use existing implementation from `searches/binary_search.py`

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

**Comment by testourii** [comment]
@dhruvmanila I used the existing implementation of searches/binary_search.py as you mentioned. To access the required file (which is in another folder (searches), I added the following lines:
```
import sys
sys.path.append("../searches")
from binary_search import bisect_left
```
which violates flake8's hooks for using an import after an instruction(E402) . By sequence, the test failed.

File: sorts/binary_insertion_sort.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Binary insertion sort is a sorting algorithm which is similar to the insertion sort,
+but instead of using linear search to find the location where an element should be
+inserted, we use binary search. Thus, we reduce the comparative value of inserting
+a single element from O (N) to O (log N).
+
+For doctests run following command:
+python3 -m doctest -v binary_insertion_sort.py
+
+For manual testing run:
+python3 binary_insertion_sort.py
```
Comment (dhruvmanila) [inline sorts/binary_insertion_sort.py:13]: Please update this part as per the recent changes.

**Comment by testourii** [comment]
@dhruvmanila  I think I fixed the issue and the docs. Could you take a look?

**Comment by testourii** [comment]
@dhruvmanila  i think the branch is ready to be merged

**Review by tianyizheng02 — COMMENTED** [review]
Thanks for your contribution, but we already have an implementation of binary insertion sort that's very similar to yours. I know that your PR is older than the current implementation and should've probably been merged first, but unfortunately the other maintainers merged a newer PR instead—sorry.

**Pull request closed without merging**